### PR TITLE
Add log export benchmark to measure the cost paid for async abstraction

### PR DESCRIPTION
**Pull request by cijothomas**
The current export method is marked async. Even when scenarios where export is done synchronously, the async is causing some unnecessary overheads (likely related to creating/updating the async state machines etc...)
This PR just shows the overhead is ~10% (250 ns ->275 ns). Not sure if this is mimicking the actual exporter correctly, would love to get some feedback on this.

### CODE REVIEW

File: opentelemetry-sdk/benches/log_exporter.rs [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+        ExportingProcessorWithFuture {
+            exporter: Mutex::new(exporter),
+        }
+    }
+}
+
+impl LogProcessor for ExportingProcessorWithFuture {
+    fn emit(&self, data: &mut LogData) {
+        let mut exporter = self.exporter.lock().expect("lock error");
+        futures_executor::block_on(exporter.export(vec![data.clone()]));
```
Comment (TommyCpp) [inline opentelemetry-sdk/benches/log_exporter.rs:68]: Do we block on runtime our LogProcrssor? It's generally not really performat

Comment (TommyCpp) [reply opentelemetry-sdk/benches/log_exporter.rs:68]: We could try https://docs.rs/tokio/latest/tokio/runtime/struct.Runtime.html#method.block_on and see if there is difference. Maybe spin up the runtime as part of the Processor so we can reuse them

**Comment by cijothomas** [comment]
> So this is measuring the case when the user doesn't have an async runtime?

Its immaterial if user has async runtime or not. The exporter (eg: exporters to systems like Windows ETW, Linux user_events) does not need async runtime, as it is executing synchronously.

**Comment by TommyCpp** [comment]
> So

OK so it's testing the case where the exporter doesn't need async runtime. Does it covers the case where the exporter requires async runtime. e.g HTTP or gRPC exporter based on async HTTP and gRPC crate?

**Review by TommyCpp — COMMENTED** [review]
> likely related to creating/updating the async state machines etc.

State machine should create during compiling time. Given the async function doesn't do anything the state machine probably only have one state. I think the overhead is mostly coming from the runtime setup here rather than the state machine itself.

**Comment by cijothomas** [comment]
> > Its immaterial if user has async runtime or not. The exporter (eg: exporters to systems like Windows ETW, Linux user_events) does not need async runtime, as it is executing synchronously.
> 
> OK so it's testing the case where the exporter doesn't need async runtime. Does it covers the case where the exporter requires async runtime. e.g HTTP or gRPC exporter based on async HTTP and gRPC crate?

It doesn't.

File: opentelemetry-sdk/benches/log_exporter.rs (hunk 1 above)
Comment (lalitb) [reply opentelemetry-sdk/benches/log_exporter.rs:68]: +1. This is not testing the async runtime. Better to create the async runtime outside of the timing loop, and then block using tokio runtime.

Comment (cijothomas) [reply opentelemetry-sdk/benches/log_exporter.rs:68]: Modified to use Tokio runtime's `block_on`, and  perf drastically dropped. (280ns to 350 ns)
```
#[derive(Debug)]
struct ExportingProcessorWithFuture {
    exporter: Mutex<NoOpExporterWithFuture>,
    rt: Runtime,
}

impl ExportingProcessorWithFuture {
    fn new(exporter: NoOpExporterWithFuture) -> Self {
        let rt = Runtime::new().unwrap();
        ExportingProcessorWithFuture {
            exporter: Mutex::new(exporter),
            rt,      
        }
    }
}

impl LogProcessor for ExportingProcessorWithFuture {
    fn emit(&self, data: &mut LogData) {
        let mut exporter = self.exporter.lock().expect("lock error");
        // futures_executor::block_on(exporter.export(vec![data.clone()]));
        self.rt.block_on(exporter.export(vec![data.clone()]));
    }

    fn force_flush(&self) -> LogResult<()> {
        Ok(())
    }

    fn shutdown(&self) -> LogResult<()> {
        Ok(())
    }
}
```

**Comment by TommyCpp** [comment]
> There is no runtime here

`futures_executor::block_on` is a runtime(well, maybe `async executor` will be a better name) as async tasks doesn't run without a execturor

> [All asynchronous computation occurs within an executor](https://docs.rs/futures-executor/latest/futures_executor/)

It doesn't change tha fact async tasks has overhead because of the need of exector

File: opentelemetry-sdk/benches/log_exporter.rs [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+#[derive(Debug)]
+struct NoOpExporterWithoutFuture {}
+impl LogExporterWithoutFuture for NoOpExporterWithoutFuture {
+    fn export(&mut self, _batch: Vec<LogData>) {}
+}
+
+#[derive(Debug)]
+struct ExportingProcessorWithFuture {
+    exporter: Mutex<NoOpExporterWithFuture>,
+}
```
[the remark this answers is not in the document]
Comment (cijothomas) [reply opentelemetry-sdk/benches/log_exporter.rs:55]: want planning to merge this as is. I started with a simpler example then slowly moved to match log exporter scenario and ended up with this. 😂
The whole point was to validate if async has overhead or not based on the last sig discussion.

**Comment by TommyCpp** [comment]
> The discussion about runtime etc. is not very relevant (apologies my knowledge of async runtime is very limited).

Yep. Sorry I kind of hijacked the discussion 😞 


> I have some workarounds

Would it make sense to offer two versions(one for sync, one for async). It's not uncommon in Rust eco system. One example would be `reqwest`, which offers `async` and [blokcing](https://docs.rs/reqwest/latest/reqwest/blocking/index.html) implementation

**Review by TommyCpp — APPROVED** [review]
Benchmark looks good. I have a few thoughts on how to setup runtime/try differnt runtime to minimize the cost. I can explore it offline once we merge it

**Review by lalitb — APPROVED** [review]

**Comment by cijothomas** [comment]
Opened an issue to track the change : https://github.com/open-telemetry/opentelemetry-rust/issues/2031
Wasn't planning to merge this benchmark originally, but no harm in keeping it either.

**Pull request merged**